### PR TITLE
Add missing console commands for client management

### LINKED ISSUE #3: Add console commands for client management

### PR DESCRIPTION
**Pull request by Allypost**
Fix #3

### CODE REVIEW

File: Command/ListClientsCommand.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Trikoder\Bundle\OAuth2Bundle\Manager\ClientFilter;
+use Trikoder\Bundle\OAuth2Bundle\Manager\ClientManagerInterface;
+use Trikoder\Bundle\OAuth2Bundle\Model\Client;
+
+final class ListClientsCommand extends Command
+{
+    public static $allowedColumns = ['identifier', 'secret', 'scope', 'redirect uri', 'grant type'];
```
Comment (spideyfusion) [inline Command/ListClientsCommand.php]: @Allypost Make it a `const`.

File: Command/ListClientsCommand.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    protected static $defaultName = 'trikoder:oauth2:list-clients';
+    private $entityManager;
```
Comment (spideyfusion) [inline Command/ListClientsCommand.php]: @Allypost Now it's called `clientManager`.

File: Manager/ClientFilter.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace Trikoder\Bundle\OAuth2Bundle\Manager;
+
+final class ClientFilter
+{
+    private $grants = null;
+    private $redirectUris = null;
+    private $scopes = null;
+
+    public static function createFilter(): self
```
Comment (spideyfusion) [inline Manager/ClientFilter.php]: @Allypost Could be just `create`.

File: Manager/ClientFilter.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        return new static();
+    }
+
+    /**
+     * @param string|array $grantOrGrants
+     */
+    public function addGrantCriteria($grantOrGrants): self
+    {
+        if (!isset($grantOrGrants)) {
```
Comment (spideyfusion) [inline Manager/ClientFilter.php]: @Allypost Check for `null` explicitly everywhere where you are using `isset`.

File: Manager/ClientManagerInterface.php [hunk 5]
```diff
@@ -9,4 +9,11 @@ interface ClientManagerInterface
     public function find(string $identifier): ?Client;
 
     public function save(Client $client): void;
+
+    public function remove(Client $client): void;
+
+    /*
+     * @return Client[]
+     */
+    public function list(?ClientFilter $filterBy): array;
```
Comment (spideyfusion) [inline Manager/ClientManagerInterface.php]: @Allypost Should the argument now be called `filter`?

File: Manager/Doctrine/ClientManager.php [hunk 6]
```diff
@@ -34,4 +35,51 @@ public function save(Client $client): void
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function list(?ClientFilter $clientFilter): array
+    {
+        $repository = $this->entityManager->getRepository(Client::class);
+
+        if (!$clientFilter || !$clientFilter->hasFilters()) {
+            return $repository->findAll();
```
Comment (spideyfusion) [inline Manager/Doctrine/ClientManager.php]: @Allypost Check if you can simplify your flow just by calling the `findBy` method with an empty array.

File: Manager/InMemory/ClientManager.php [hunk 7]
```diff
@@ -27,4 +28,49 @@ public function save(Client $client): void
...
+        };
+    }
+
+    private static function passesFilter(array $clientValues, ?array $filterValues): bool
+    {
+        if (!$filterValues) {
+            return true;
+        }
+
+        $clientGrants = array_map(self::stringifyFunction(), $clientValues);
```
Comment (spideyfusion) [inline Manager/InMemory/ClientManager.php]: @Allypost You can just use `strval` here as a callback.

File: Manager/InMemory/ClientManager.php [hunk 8]
```diff
@@ -27,4 +28,49 @@ public function save(Client $client): void
...
+        $grantsPassed = array_intersect($filterValues, $clientGrants);
```
Comment (spideyfusion) [inline Manager/InMemory/ClientManager.php]: @Allypost Naming. :)

File: Tests/Acceptance/ListClientsCommandTest.php [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace Trikoder\Bundle\OAuth2Bundle\Tests\Acceptance;
+
+use Symfony\Component\Console\Tester\CommandTester;
+use Trikoder\Bundle\OAuth2Bundle\Manager\ClientManagerInterface;
+use Trikoder\Bundle\OAuth2Bundle\Model\Client;
+use Trikoder\Bundle\OAuth2Bundle\Model\RedirectUri;
+use Trikoder\Bundle\OAuth2Bundle\Model\Scope;
+
+final class ListClientsCommandTest extends AbstractAcceptanceTest
```
Comment (spideyfusion) [inline Tests/Acceptance/ListClientsCommandTest.php]: @Allypost Add a test case for when the client list is empty.

File: Manager/ClientManagerInterface.php [hunk 10]
```diff
@@ -9,4 +9,11 @@ interface ClientManagerInterface
     public function find(string $identifier): ?Client;
 
     public function save(Client $client): void;
+
+    public function remove(Client $client): void;
```
Comment (X-Coder264) [inline Manager/ClientManagerInterface.php]: @Allypost @spideyfusion Adding methods to an interface is a breaking change and as such this PR can be merged only for the next major version which I suppose will be tagged soon after https://github.com/trikoder/oauth2-bundle/pull/43 gets merged.

Comment (Allypost) [reply Manager/ClientManagerInterface.php]: https://semver.org/#spec-item-7
This change is backwards compatible functionality. Nothing backwards incompatible was added/changed.

File: Manager/ClientFilter.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+    public static function create(): self
+    {
+        return new static();
+    }
+
+    /**
+     * @param string|array|null $grantOrGrants
+     */
+    public function addGrantCriteria($grantOrGrants): self
```
[the remark this answers is not in the document]
Comment (Allypost) [reply Manager/ClientFilter.php]: @HypeMC I left it at just setting it as a variadic function. I think new methods should be added to remove criteria.

File: Command/ListClientsCommand.php [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    protected function configure(): void
+    {
+        $this
+            ->setDescription('Lists existing oAuth2 clients')
+            ->addOption(
+                'columns',
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'Determine which columns are shown. Comma separated list.',
+                implode(', ', self::ALLOWED_COLUMNS)
```
[the remark this answers is not in the document]
Comment (Allypost) [reply Command/ListClientsCommand.php]: @HypeMC They're all required arrays with default values

File: Command/ListClientsCommand.php [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setDescription('Lists existing oAuth2 clients')
+            ->addOption(
+                'columns',
+                null,
+                InputOption::VALUE_OPTIONAL,
```
[the remark this answers is not in the document]
Comment (Allypost) [reply Command/ListClientsCommand.php]: @HypeMC They're all required arrays with default values now

File: Manager/ClientManagerInterface.php (hunk 10 above)
Comment (Allypost) [reply Manager/ClientManagerInterface.php]: @X-Coder264 Yeah this will get merged after the version bump.

File: Manager/ClientManagerInterface.php [hunk 14]
```diff
@@ -9,4 +9,11 @@ interface ClientManagerInterface
...
+
+    /*
```
[the remark this answers is not in the document]
Comment (Allypost) [reply Manager/ClientManagerInterface.php]: @HypeMC Whoops! Fixed.

File: Command/ListClientsCommand.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        $io = new SymfonyStyle($input, $output);
+        $columns = $this->getColumns($input);
+        $rows = $this->getRows($clients, $columns);
+        $io->table($columns, $rows);
+    }
+
+    private function getRows(array $clients, array $columns): array
+    {
+        return array_map(function (Client $client) use ($columns) {
```
[the remark this answers is not in the document]
Comment (Allypost) [reply Command/ListClientsCommand.php]: @HypeMC I added type hints to everything now :)

File: Tests/Acceptance/DeleteClientCommandTest.php [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace Trikoder\Bundle\OAuth2Bundle\Tests\Acceptance;
+
+use Symfony\Component\Console\Tester\CommandTester;
+use Trikoder\Bundle\OAuth2Bundle\Manager\ClientManagerInterface;
+use Trikoder\Bundle\OAuth2Bundle\Model\Client;
+
+final class DeleteClientCommandTest extends AbstractAcceptanceTest
+{
+    public function testDeleteClient()
```
[the remark this answers is not in the document]
Comment (Allypost) [reply Tests/Acceptance/DeleteClientCommandTest.php]: @HypeMC I just went ahead and added them to everything. Can't hurt.

File: Manager/ClientManagerInterface.php (hunk 10 above)
Comment (X-Coder264) [reply Manager/ClientManagerInterface.php]: @Allypost My point was this should be merged before the next major version gets tagged. Anyway, this discussion isn't important for this PR, @spideyfusion will take care of that :)

File: Manager/ClientFilter.php [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        return $this->scopes;
+    }
+
+    public function hasFilters(): bool
+    {
+        return
+            null !== $this->grants
+            || null !== $this->redirectUris
+            || null !== $this->scopes;
+    }
```
Comment (HypeMC) [inline Manager/ClientFilter.php]: @Allypost If the filters an empty array `hasFilters` will return true, not sure how I feel about that.

File: Command/ListClientsCommand.php [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+                'grant-type',
+                null,
+                InputOption::VALUE_IS_ARRAY | InputOption::VALUE_REQUIRED,
+                'Finds by allowed grant type for client. Use this option multiple times to filter by multiple grant types.',
+                []
+            )
+            ->addOption(
+                'scope',
+                null,
+                InputOption::VALUE_IS_ARRAY | InputOption::VALUE_OPTIONAL,
```
Comment (HypeMC) [inline Command/ListClientsCommand.php]: @Allypost `InputOption::VALUE_OPTIONAL`?

File: Manager/ClientManagerInterface.php (hunk 10 above)
Comment (spideyfusion) [reply Manager/ClientManagerInterface.php]: @X-Coder264 Indeed. :)

File: Manager/ClientFilter.php (hunk 17 above)
Comment (Allypost) [reply Manager/ClientFilter.php]: @HypeMC All the values should be arrays by default now. I switched this for `empty` checks.

File: Command/ListClientsCommand.php (hunk 18 above)
Comment (Allypost) [reply Command/ListClientsCommand.php]: @HypeMC The one that got away :(
